TODO List
=========

http://www.slideshare.net/jordixou/nginx-uwsgi-emperor-bottle

Session Management
------------------

Session in s single bottlesession.py file 
http://conceptpending.com/session-in-a-bottle/

Has a session cookie generator, full example with template.
https://github.com/linsomniac/bottlesession

Basic Authentication in bottle.py
http://fledglingpolymath.tumblr.com/post/39977900894/basic-authentication-in-bottle-py

Really good collection of snippets for Flask but some apply to Bottle.
http://flask.pocoo.org/snippets/   
http://flask.pocoo.org/snippets/category/sessions/
See e.g. http://pythonhosted.org/Frozen-Flask/


Timeout Errors: April, May 2015
-------------------------------

uwsgi connect_timeout to set the gateway Timeout

https://www.tekovic.com/fixing-timeout-between-nginx-and-php-fpm
The error clearly states that connection timed out between Nginx and upstream
server - in this case PHP-FPM, but that could be any FastCGI server - while
reading response header. If you put 2 and 2 together, you just can't miss the
fastcgi_read_timeout directive. 


http://wiki.nginx.org/HttpUwsgiModule#uwsgi_read_timeout
    uwsgi_connect_timeout <-- sets timeout period for connection with uWSGI-server.
        It should be noted that this value can't exceed 75 seconds.
    uwsgi_read_timeout 

504 errors 

Current timeout Values are:
    /etc/nginx/nginx.conf:    
        keepalive_timeout  75;      <-- explicit
        uwsgi_connect_timeout  60s  <-- default
        uwsgi_read_timeout 60s;
        uwsgi_send_timeout 60s;

    # These timeouts don't help in preventing "504 Gateway Time-out"
    #keepalive_timeout  65;
    #keepalive_timeout 200;
    #uwsgi_connect_timeout 200;
    #uwsgi_send_timeout 200; 
    # This timeout is the one required!
    uwsgi_read_timeout 200; 


September 2014
---------------

Error: 500 Internal Server Error
Sorry, the requested URL 'http://130.56.248.113/choice/process/' caused an error:
Internal Server Error

This error is not logged under /var/log/nginx/choice_error.log
It's logged at /var/log/uwsgi.log

[root@uts-choice etc]# tail /var/log/uwsgi.log 
  File "/usr/lib/python2.7/site-packages/bottle.py", line 1574, in wrapper
    def set_cookie(self, name, value, secret=None, **options):
  File "/home/ec2-user/public_html/choice/choice.py", line 416, in process
    outputs = read_output_files(outputs_to_read, tempdir)
  File "/home/ec2-user/public_html/choice/choice.py", line 175, in read_output_files
    write_errors('Output file: %s (%s)\n' % (filename, e))
  File "./choice_common.py", line 27, in write_errors
    fh = open('errors.txt', 'a')
IOError: [Errno 13] Permission denied: 'errors.txt'
[pid: 27375|app: 0|req: 85/85] 138.25.35.165 () {44 vars in 768 bytes} [Fri Sep 12 15:27:19 2014] POST /choice/process/ => generated 756 bytes in 2115 msecs (HTTP/1.1 500) 2 headers in 99 bytes (1 switches on core 0)
[root@uts-choice etc]# 

Temporarily I have made an /tmp/errors.txt that is writable by uwsgi 

[root@uts-choice tmp]# cat errors.txt 
Output file: out_cinv.dat ([Errno 2] No such file or directory: '/tmp/choice.Hhl7q8/out_cinv.dat')
Output file: out_correln.dat ([Errno 2] No such file or directory: '/tmp/choice.Hhl7q8/out_correln.dat')
[root@uts-choice tmp]# 



Fix not found errors in nginx logs
----------------------------------

In nginx logs we get:

2014/05/05 10:02:02 [error] 29064#0: *80 open() "/home/ec2-user/public_html/404.html" failed (2: No such file or directory), client: 124.122.140.15, server: localhost, request: "GET /myadmin/scripts/setup.php HTTP/1.1", host: "130.56.248.113"

This is becasue in nginx/conf.d/default.conf we have:

  # redirect server error pages to the static page /40x.html
  error_page  404  /404.html;
  location = /40x.html {
  }
  
  # redirect server error pages to the static page /50x.html
  error_page  500 502 503 504  /50x.html;
  location = /50x.html {
  }


User Requests
-------------

Emily Bird in email dated 20.01.2014 asked:

1) Order of the output. Can we please have the following order:
   - B Matrix
   - Lambda Matrix
   - C Matrix
   - C Inverse
   - Correlation Matrix

DONE

2) Remove the time/date from output (but it looks like this request is out of
   date - the time/date doesn't appear in the output anymore, does it?)

DONE

3) Remove the output "Your input seems fine". 

DONE

4) When "Main effects + all two-factor interactions" or "Main effects + some
   two-factor interactions" is selected but no determinant is entered, display
   "None" under the heading "Determinant of C Entered" (at the moment it is just
   blank).


Implementing a Timeout
----------------------

Refs:
http://stackoverflow.com/questions/1191374/subprocess-with-timeout
http://stackoverflow.com/questions/3733270/python-subprocess-timeout
http://howto.pui.ch/post/37471155682/set-timeout-for-a-shell-command-in-python

See 
  p = process.ProcessOpen("notepad.exe")
  p.wait(timeout=5)


subprocess.call(): 

See pydoc subprocess for "call" details. 
This returns 0 if all is OK.
To capture standard error in the result use stderr=STDOUT.
The exception object will have one extra attribute called
'child_traceback', which is a string containing traceback information
from the childs point of view.
See also "check_call"

(['./process_choices.py', 'construct', 'main'], stderr=subprocess.STDOUT)
return_value = subprocess.call(['./process_choices.py', tempdir, 'construct', 'main'] )



